images/router/haproxy/conf/haproxy-config.template

```diff
@@ -769,6 +769,23 @@ backend {{ genBackendNamePrefix $cfg.TLSTermination }}:{{ $cfgIdx }}
               {{- else }}
                 {{- if not (isTrue $router_disable_http2) }} alpn h2,http/1.1
                 {{- end }}
+                {{- /*
+                    Add "verifyhost" with service hostname only if both are true:
+                    - The route has a single active service.
+                    - The route does not have a destination CA certificate.
+
+                    Limiting "verifyhost" to a single service route is essential,
+                    as it's not possible to reliably predict which service’s endpoint will be used
+                    when alternate backends are available.
+                */ -}}
+                {{- if and $cfg.VerifyServiceHostname (eq $cfg.ActiveServiceUnits 1) -}}
+                    {{- range $serviceUnitName, $weight := $cfg.ServiceUnitNames -}}
+                        {{- with $serviceUnit := index $.ServiceUnits $serviceUnitName -}}
+                            {{- print " verifyhost " $serviceUnit.Hostname }}
+                            {{- break }}
+                        {{- end }}
+                    {{- end }}
+                {{- end }}
```

Comment:
I don't know what our philosophy is on DCM, or how much efficiency we are trying to squeeze out of it, but did you consider the strategy of always assuming dynamic servers are reserved for the primary backend (not the alternates), and change your logic to only reload if an alternate backend endpoints needs syncing?

While it might be an asymmetrical efficiency gain, it'd be a net gain to avoid reloading for primary backend endpoint changes. Net positive, but yes, asymmetrical gains (primary wins while alternate lose). Also depends the added complexity and whether that provides reasonable ROI. Not trying to demand we need to be as perfect as possible, but I'm just curious.

Reply:
> Upd: just recalled that the main problem was to bring the notion of alternate service to the template, all service units of a route are the same for the moment.

I see, it'd require more refactoring than probably it's worth for the added efficiency since we don't distinguish between the primary backend and alternate backends. It's something we can keep in mind for later improvements.

Reply:
[This commit](https://github.com/openshift/router/pull/637/commits/17d42d1cc1c16523420bf6c98d7932545e3d9bc5) can do what you proposed.

Reply:
:+1: Looks good, doesn't seem to add too much complexity, and it made the fix slightly more efficient.